### ISSUE TITLE
Remove the `class_name` keyword and use only `class`

### ISSUE THREAD

**Issue by dmitryuck** [godotengine/godot-proposals#1842]
**Describe the project you are working on:**
Not important

**Describe the problem or limitation you are having in your project:**
for declaring the name of the class we use "class_name MyClass"
but for declaring internal classes keyword "class InternalClass"

**Describe the feature/enhancement and how it helps to overcome the problem or limitation:**
the idea is to remove "class_name" and use only "class"
By doing this we can throw out one keyword

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
If we want to set a name for a class we can check if there is no class body declaration then it is setting a name for the class

class MyClass #this obviously that it is giving a name to the class

class InternalClass:
  func MyFunc:
   pass

this is an internal class declaration so no confusing

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
no

**Is there a reason why this should be core and not an add-on in the asset library?:**
related to GDScript syntax

**Comment by katoneko** [comment]
A very similar thing was discussed in #209. The consensus was that changing the current way for the sake of changing has little actual benefit. And using `class` instead of `class_name` would introduce confusion with the declaration of inner classes as per [this comment](https://github.com/godotengine/godot-proposals/issues/209#issuecomment-624193796).

**Comment by pycbouh** [comment]
> By doing this we can throw out one keyword

That's not a benefit in and of itself, though. Different keywords are good. Different keywords make code less ambiguous. We shouldn't aim to make the language as concise as possible. We should aim to make it as clear as possible.

**Comment by katoneko** [comment]
One thing I see that could be discussed here that wasn't covered in #209 is removing the underscore in `class_name` and making it `classname`. We already have `onready` without underscores, so the change would make it more consistent.

**Comment by dmitryuck** [comment]
icon 'game/icons/icon.ico' could be added as @icon metadata

@icon='game/icons/icon.ico'
class MyNode extends Node

internal classes provided in the old way
class InternalClass

**Comment by Xrayez** [comment]
> A very similar thing was discussed in #209. The consensus was that changing the current way for the sake of changing has little actual benefit. And using `class` instead of `class_name` would introduce confusion with the declaration of inner classes as per [this comment](https://github.com/godotengine/godot-proposals/issues/209#issuecomment-624193796).

There are actually cases where `class_name` creates problems with existing method names/properties: godotengine/godot#43255. Of course, those can be renamed to something else, but by using `class` those kind of errors can be eliminated completely, because it's unlikely that someone would want to bind a property with `class` as identifier due to how C++ works. Also lets not forget about C++ modules which may be cursed by limitations like this.

Also, https://github.com/godotengine/godot-proposals/issues/209#issuecomment-624193796 only describes that it's "more difficult to implement" (it's just a matter of work) and that it would somehow "create confusion" with inner classes. I've never had to declare a class with other nested inner classes so I consider this as corner case, I'd rather prefer to have consistency in syntax for declaring classes, even if this requires some parsing nuances to be overcome.

Again, the fact that this proposal is created which is basically #209 only proves that there's indeed a problem.

**Comment by pycbouh** [comment]
@Xrayez This particular proposal didn't even describe any problem, though, just a random wish.

As far as confusion, if that's vnen you are quoting, he was probably talking about confusing logic in the parser, as this change would make the class keyword ambiguous.

**Comment by DriNeo** [comment]
Is it necessary for the gdscript to have unnamed classes ?
Why not automatically assign the name of the file ?

**Comment by pycbouh** [comment]
`class_name` doesn't just name a script, it registers it in global scope. Not every script needs to be globally accessible. Most don't.

**Comment by Shadowblitz16** [comment]
I really agree class_name is confusing and repetitive.

we already have a export keyword and annotation lets use it.
I also want to see the ability to have multiple nodes defined per script

```
export class MyCustomNode extends Node2D:
    func _ready():
        pass

export class MyCustomNode2 extends Spatial:
    func _ready():
        pass
```

this would work well with #615

**Comment by aspenforest** [comment]
> One thing I see that could be discussed here that wasn't covered in #209 is removing the underscore in `class_name` and making it `classname`. We already have `onready`, `remotesync`, `mastersync`, `puppetsync` without underscores, so the change would make it more consistent.

I dislike the underscore of class_name, because class_name is the single keyword with an underscore and thus its incongruent with the general keyword naming scheme.

**Comment by Shadowblitz16** [comment]
wall its good practice to name your classes so you can actually use them in the first place
but underscore could be a option

**Comment by LennyPhoenix** [comment]
> I also want to see the ability to have multiple nodes defined per script

How would you select which class to use when assigning the script to a node?

**Comment by Shadowblitz16** [comment]
It would use the first one.

**Comment by Shadowblitz16** [comment]
@LennyPhoenix 
Maybe it could load all scripts with the correct type or there could be a selection dialog

**Comment by SysError99** [comment]
I agree with @katoneko , `classname` is much more appropriate. At least makes it as main keyword and drop `class_name` in the future version of Godot.

**Comment by bbugh** [comment]
Apologies if this has been mentioned before, I didn't see this mentioned in any of the threads leading to this one.

Is there a particular technology concern with having the class name default to the file name as  PascalCase? I ran into an issue where one custom class couldn't access another custom class, and it took a while to figure out that I needed to define `class_name Thing`, since it seemed like it would be automatic like other languages.

**Comment by Calinou** [comment]
> Is there a particular technology concern with having the class name default to the file name as PascalCase?

I'm not sure if implicitly depending on the file name is a good idea. It could break things when users copy-paste scripts across projects and rename them.

**Comment by MichaelGlenMontague** [comment]
Definitely prefer `classname` over `class_name`. It bugs me how it's currently the only GDScript keyword that has an underscore in between words. Not including the shading language, of course.

**Comment by Calinou** [comment]
I like the idea of using `classname` over `class_name`, especially since we already have compound words as keywords such as `onready` (`@onready` annotation in 4.0).

**Comment by vnen** [comment]
I don't mind changing to `classname`, it's an easy fix. The question is really whether it's worth breaking compatibility just for the sake of changing it. I know GDScript is already different, but it broke only what was really needed.

We need to settle this soon since the window for 4.0 is closing and after that we won't do any change.

**Comment by aspenforest** [comment]
> I prefer `class_name` than `classname`, if all lower, I think should use underline, If not underline, I like `ClassName`

While _under_score_ identifiers are the best practice for Python variable names and CamelCase for Python class names, nowhere in Python are these naming conventions used for reserved **_keywords_**, which the class_name word in GDScript is. Yes, I know that GDScript is not Python, but it is heavily influenced by Python, therefore, would it not make sense to use a compatible naming convention, where the naming convention of variables is not used for reserved keywords of the language?

**Comment by bbugh** [comment]
Like some others have said, I think if it's going to change giving it a more precise name would improve the usage. Some off-the-cuff variations that don't require underscores and are more precise:

```gdscript
public SomeClass
publish SomeClass
publish as SomeClass
export SomeClass
export as SomeClass
generate SomeClass
declare SomeClass
```

Personally I like `export` and `declare`, though I know `export` is already used for something much different.

**Comment by aaronfranke** [comment]
I'm still in favor of renaming this to `class`, and if necessary, change inner classes to a different keyword. We could even go further and make the name of the class implicit. See the discussion in #209.

**Comment by Shadowblitz16** [comment]
Why can't we just have class and get rid of class_name?
Annotations could be used to export it to the node creation dialog?

And as for what script is used for the node, godot can just do the same thing it does for C#.

class_name is kinda a gd script only thing anyways, even python would probably use a class, 
Also getting rid of class_name you guys can work on improving just class.

**Comment by aaronfranke** [comment]
@Shadowblitz16 `class_name` does more than export it to the node creation dialog. It also makes it work as a type in GDScript. So you can write `MyClass.new()` to make an instance instead of `preload("my_class.gd").new()`.

**Comment by Shadowblitz16** [comment]
> @Shadowblitz16 `class_name` does more than export it to the node creation dialog. It also makes it work as a type in GDScript. So you can write `MyClass.new()` to make an instance instead of `preload("my_class.gd").new()`.

why can't class do this too?
why do we need `MyClass.new()` and not just `MyClass()`?

**Comment by aaronfranke** [comment]
@Shadowblitz16 Having instances of classes created without `.new` like we have with vectors is a completely unrelated proposal. I'm not sure if that's desired, but if you want to discuss it, it should go in a new proposal.

**Comment by Shadowblitz16** [comment]
> @Shadowblitz16 Having instances of classes created without `.new` like we have with vectors is a completely unrelated proposal. I'm not sure if that's desired, but if you want to discuss it, it should go in a new proposal.

You just said it was related, now it's not? I don't understand.

**Comment by aaronfranke** [comment]
@Shadowblitz16 Look closer. The `.new()` was an example, the point was `preload("my_class.gd")` -> `MyClass`.

**Comment by Shadowblitz16** [comment]
> @Shadowblitz16 Look closer. The `.new()` was an example, the point was `preload("my_class.gd")` -> `MyClass`.

oh I see why can't we have a annotation for that as well?
class should be registering type names anyways.

**Comment by akien-mga** [comment]
After discussion, we decided not to change `class_name`.

Reusing the `class` keyword for it as initially proposed is problematic as it's not the same as declaring an inner class with `class Name:`. It would basically require having two different keywords using the same name.

Changing the name to `classname` to remove the underscore, after a lot of bikeshedding, was not seen as particularly wanted. There's no strong opposition but also no strong support, and so it doesn't seem useful to break compatibility here.

We also discussed making it an annotation, but this was also rejected as discussed here: https://github.com/godotengine/godot/pull/59698#issuecomment-1087691988
